deepchem/data/datasets.py

```diff
@@ -636,65 +637,127 @@ def iterbatches(self,
                   epoch=0,
                   deterministic=False,
                   pad_batches=False):
-    """Get an object that iterates over minibatches from the dataset.
-
+    """ Get an object that iterates over minibatches from the dataset. It is guaranteed
+    that the number of batches returned is math.ceil(len(dataset)/batch_size).
+    
     Each minibatch is returned as a tuple of four numpy arrays: (X, y, w, ids).
+
+
+    Parameters:
+    -----------
+    batch_size: int
+      Number of elements in a batch. If None, then it yields batches with size equal to the size
+      of each individual shard.
+
+    epoch: int
+      Not used
+
+    deterministic: bool
+      Whether or not we should should shuffle each shard before generating the batches.
+      Note that this is only local in the sense that it does not ever mix between different
+      shards.
+
+    pad_batches: bool
+      Whether or not we should pad the last batch, globally, such that it has exactly batch_size
+      elements.
+
+
     """
 
-    def iterate(dataset):
+    def iterate(dataset, batch_size):
       num_shards = dataset.get_number_shards()
       if not deterministic:
         shard_perm = np.random.permutation(num_shards)
       else:
         shard_perm = np.arange(num_shards)
-      pool = Pool(1)
+
+      # (ytz): Depending on the application, thread-based pools may be faster
+      # than process based pools, since process based pools need to pickle/serialize
+      # objects as an extra overhead. Also, as hideously as un-thread safe this looks,
+      # we're actually protected by the GIL.
+      pool = Pool(1)  # mp.dummy aliases ThreadPool to Pool
       next_shard = pool.apply_async(dataset.get_shard, (shard_perm[0],))
-      for i in range(num_shards):
+
+      total_yield = 0
+
+      if batch_size is None:
+        num_global_batches = num_shards
+      else:
+        num_global_batches = math.ceil(len(dataset) / batch_size)
+
+      cur_global_batch = 0
+      cur_shard = 0
+      carry = None
+
+      while cur_global_batch < num_global_batches:
+
         X, y, w, ids = next_shard.get()
-        if i < num_shards - 1:
-          next_shard = pool.apply_async(dataset.get_shard, (shard_perm[i + 1],))
-        n_samples = X.shape[0]
-        # TODO(rbharath): This happens in tests sometimes, but don't understand why?
-        # Handle edge case.
-        if n_samples == 0:
-          continue
-        if not deterministic:
-          sample_perm = np.random.permutation(n_samples)
+        if cur_shard < num_shards - 1:
+          next_shard = pool.apply_async(dataset.get_shard,
+                                        (shard_perm[cur_shard + 1],))
         else:
-          sample_perm = np.arange(n_samples)
+          pool.close()
+
+        if carry is not None:
+          X = np.concatenate([carry[0], X], axis=0)
+          y = np.concatenate([carry[1], y], axis=0)
+          w = np.concatenate([carry[2], w], axis=0)
```

Comment:
do we have to do None checks here?
there are cases when y and w are None.

Reply:
@lilleswing you're right about this. Will add a guard.

Reply:
Oddly enough even pad_batches doesn't guard against this:

``` python
def pad_batch(batch_size, X_b, y_b, w_b, ids_b):
  """Pads batch to have size precisely batch_size elements.

  Fills in batch by wrapping around samples till whole batch is filled.
  """
  num_samples = len(X_b)
  if num_samples == batch_size:
    return (X_b, y_b, w_b, ids_b)
  # By invariant of when this is called, can assume num_samples > 0
  # and num_samples < batch_size
  if len(X_b.shape) > 1:
    feature_shape = X_b.shape[1:]
    X_out = np.zeros((batch_size,) + feature_shape, dtype=X_b.dtype)
  else:
    X_out = np.zeros((batch_size,), dtype=X_b.dtype)

  num_tasks = y_b.shape[1] # fucked if None
  y_out = np.zeros((batch_size, num_tasks), dtype=y_b.dtype)
  w_out = np.zeros((batch_size, num_tasks), dtype=w_b.dtype) # fucked if None
  ids_out = np.zeros((batch_size,), dtype=ids_b.dtype)

  # Fill in batch arrays
  start = 0
  # Only the first set of copy will be counted in training loss
  w_out[start:start + num_samples] = w_b[:]
  while start < batch_size:
    num_left = batch_size - start
    if num_left < num_samples:
      increment = num_left
    else:
      increment = num_samples
    X_out[start:start + increment] = X_b[:increment]
    y_out[start:start + increment] = y_b[:increment]
    ids_out[start:start + increment] = ids_b[:increment]
    start += increment

  return (X_out, y_out, w_out, ids_out)
```

Reply:
@lilleswing Is there ever a use case where pad_batch=True and y/w could be None?

Reply:
I don't think they're related.  If the dataset doesn't include weights, that's a property of the dataset.  `pad_batches` is an option determined by the calling code.